components/script/docs/JS-Servos-only-GC.md

```diff
@@ -0,0 +1,396 @@
+
+[deref]: http://doc.rust-lang.org/std/ops/trait.Deref.html
+
+The DOM methods of `Window` (for example) are defined in a trait which is
+[implemented][windowmethods] for `JSRef<Window>`. This ensures that the
+`self` pointer is rooted for the duration of the method call, which would not
+be guaranteed if we implemented the methods on `Window` directly.
+
+XXX: this is not really true, is it?
+
+[windowmethods]: XXX
+
+You can check out the [`js` module's documentation][js-docs] for more details
+that didn't make it into this document.
+
+[js-docs]: http://doc.servo.org/script/dom/bindings/js/index.html
+
+Custom static analysis
+======================
+
+To recap, the safety of our system depends on two major parts:
+
+- The auto-generated `trace` methods ensure that SpiderMonkey's garbage
+  collector can see all of the references between DOM objects.
+- The implementation of `Root<T>` and `JSRef<T>` guarantees that we can't use a
+  DOM object from Rust without telling SpiderMonkey about our temporary
+  reference.
+
+But there's a hole in this scheme. We could copy an unrooted pointer — a
+`JS<T>` — to a local variable on the stack, and then at some later point, root
+it and use the DOM object. In the meantime, SpiderMonkey's garbage collector
+won't know about that `JS<T>` on the stack, so it might free the DOM object.
+To really be safe, we need to make sure that `JS<T>` *only* appears in places
+where it will be traced, such as DOM structs, and never in local variables,
+function arguments, and so forth.
+
+// XXX: this gives me a crazy idea: what if we made JS<T> !Copy?
+
+This rule doesn't correspond to anything that already exists in Rust's type
+system. Fortunately, the Rust compiler can load 'lint plugins' providing custom
+static analysis. These basically take the form of new compiler warnings,
+although in this case we set the default severity to 'error'.
+
+[lints]: https://github.com/rust-lang/rust/pull/15024 (better link?)
+
+We have already [implemented a plugin][js-lint] which simply forbids `JS<T>`
+from appearing at all. Because lint plugins are part of the usual
+[warnings infrastructure][warnings], we can use the `allow` attribute in places
+where it's okay to use `JS<T>`, like DOM struct definitions and the
+implementation of `JS<T>` itself.
+
+[js-lint]: https://github.com/kmcallister/servo/commit/c20b50bbbbcdc8ce3551adbc1e039a727cf89995 (better link?)
+[warnings]: http://doc.rust-lang.org/rust.html#lint-check-attributes (better link?)
+
+Our plugin looks at every place where the code mentions a type. Remarkably,
+this adds only a fraction of a second to the compile time for Servo's largest
+subcomponent, as Rust compile times are dominated by [LLVM][llvm]'s back-end
+optimizations and code generation. The current version of the plugin is very
+simple and will miss some mistakes, like storing a struct containing `JS<T>` on
+the stack. (XXX does it still?) However, lint plugins run at a late stage of
```

Comment:
Such a struct can't be defined without being `must_root`, _unless_ it has generics, in which case there are ways of sidestepping the lint.